File: src/cpp/gradientDescent.cpp
#include <stdio.h>
#include <stdlib.h>
#include <fstream>

using namespace std;

double xSample[25000][385];
double ySample[25000];

class GradientDescent {
private:
    double theta[385];
    double alpha;
    int sampleCount;
    int dimension;

    void loadSample() {
        ifstream inputFile("./data/train.txt");
        int temp;
        for (int i = 0 ; i < 25000 ; i++) {
            xSample[i][0] = 1;
            inputFile >> temp;
            for (int j = 1 ; j < 385 ; j++) {
                inputFile >> xSample[i][j];
            }
            inputFile >> ySample[i];
        }
    }

    double hTheta(int index) {
        double sum = 0;
        for (int i = 0 ; i < 385 ; i++) {
            sum += theta[i] * xSample[index][i];
        }
        return sum;
    }

    double jTheta() {
        double sum = 0;
        for (int i = 0 ; i < 25000 ; i++) {
            sum += (hTheta(i) - ySample[i]) * (hTheta(i) - ySample[i]);
        }
        return sum / 2;
    }

public:
    GradientDescent(int dimen) {
        dimension = dimen;
        for (int i = 0 ; i < dimension ; i++) {
            theta[i] = 0;
        }
        alpha = 0.0001;
        sampleCount = 0;
    }
    void training() {
        loadSample();

        int step = 0;

        double lastJTheta = 0;
        double currentJTheta = 0;
        double lastTheta[385];

        while (true) {
            step++;
            if (step > 10000) {
                break;
            }

            lastJTheta = currentJTheta;

            for (int i = 0 ; i < 385 ; i++) {
                lastTheta[i] = theta[i];
            }

            for (int j = 0 ; j < 385 ; j++) {
                double tempSum = 0;
                for (int i = 0 ; i < 25000 ; i++) {
                    tempSum += (ySample[i] - hTheta(i)) * xSample[i][j];
                    theta[j] += alpha * tempSum / 25000.0;
                }
            }

            currentJTheta = jTheta();
            printf("%d: %lf\n", step, currentJTheta);

            if (lastJTheta != 0 && lastJTheta <= currentJTheta) {
                for (int i = 0 ; i < 385 ; i++) {
                    theta[i] = lastTheta[i];
                    printf("%lf ", theta[i]);
                }
                break;
            }

        }
    }

    void predicting() {
        ifstream inputFile("./data/test.txt");
        ofstream outputFile("./data/outC.csv");
        outputFile << "id,reference\n";
        for (int i = 0 ; i < 25000 ; i++) {
            int id;
            double x = 0;
            double reference = theta[0];
            inputFile >> id;
            for (int j = 1 ; j < 385 ; j++) {
                inputFile >> x;
                reference += x * theta[j];
            }
            outputFile << id << "," << reference << "\n";
        }
        outputFile.close();
    }

};


int main() {
    GradientDescent gradientDescent = GradientDescent(385);
    gradientDescent.training();
    gradientDescent.predicting();
    return 0;
}

File: src/python/gradientDescent.py
#coding=utf-8

import os

class GradientDescent:
    def __init__(self, dimension):
        self.theta_list = [0] * dimension
        self.x_sample = []
        self.y_sample = []
        self.dimension = dimension
        self.alpha = 0.01
        self.sample_count = 0

    def load_sample(self, filename):
        with open(filename) as file:
            first_line = file.readline()
            k = 0
            for line in file.readlines():
                k += 1
                # if k > 3000:
                #     break
                line = line.strip()
                line_list = map(lambda n: float(n), line.split(','))
                self.x_sample.append([1] + line_list[1:self.dimension])
                self.y_sample.append(line_list[self.dimension])
                self.sample_count += 1

    def h_theta(self, x_i):
        return sum([x * y for x, y in zip(self.theta_list, x_i)])

    def j_theta(self):
        return sum([(self.h_theta(x) - y) ** 2 for x, y in zip(self.x_sample, self.y_sample)]) / 2

    def training(self, filename):
        self.load_sample(filename)

        last_j_theta = None
        current_j_theta = None
        training_step = 0
        last_theta_list = None

        while True:
            #print training_step, last_j_theta, current_j_theta
            if training_step > 10000:
                break
            last_theta_list = self.theta_list
            last_j_theta = current_j_theta
            training_step += 1
            for j in range(0, self.dimension):
                #print j
                temp = sum([(y - self.h_theta(x)) * x[j] for x, y in zip(self.x_sample, self.y_sample)])
                self.theta_list[j] += self.alpha * temp / float(self.sample_count)

            current_j_theta = self.j_theta()
            print current_j_theta
            if last_j_theta == None:
                last_j_theta = current_j_theta
            elif abs(current_j_theta - last_j_theta) <= 1:
                self.theta_list = last_theta_list
                print self.theta_list
                break

    def predicting(self, filename):
        test_file = open(filename)
        output_file = open("./data/outputPython", "w")
        try:
            first_line = test_file.readline()
            output_file.write("id,reference\n")
            for line in test_file.readlines():
                line = line.strip()
                line_list = map(lambda n: float(n), line.split(','))
                x_sample = [1] + line_list[1:self.dimension]
                output_file.write("{0},{1}\n".format(str(int(line_list[0])), str(sum([x * y for x, y in zip(self.theta_list, x_sample)]))))
        finally:
            test_file.close()
            output_file.close()


if __name__ == '__main__':
    gradient_descent = GradientDescent(dimension=385)
    gradient_descent.training("./data/train.csv")
    gradient_descent.predicting("./data/test.csv")
